Properly handle latin-1 encoding in file diffs

terminalmage opened the pull request:
This changes the encoding/decoding helpers to attempt latin-1 after initially trying UTF-8, which prevents a traceback when displaying diffs when a file with latin-1 encoding is updated using a file.managed state.

Resolves #48777

terminalmage committed:
Add get_diff to salt.utils.stringutils

terminalmage committed:
Make to_unicode/to_str/to_bytes helpers attempt latin-1

Also allow for multiple encodings to be passed

terminalmage committed:
Use new get_diff helper in file module

terminalmage committed:
Use BASE_FILES instead of redundant STATE_DIR

terminalmage committed:
Add integration test for latin-1 file diffs

terminalmage committed:
PY3 scoping fix

In PY3 the caught exceptions now drop out of scope when leaving the for
loop.

terminalmage committed:
Add unit test for latin-1 fallback, multi-encoding

rallytime commented on salt/modules/file.py:
```diff
@@ -1567,7 +1567,7 @@ def comment_line(path,
 
     # Return a diff using the two dictionaries
-    return ''.join(difflib.unified_diff(orig_file, new_file))
+    return __salt__['stringutils.get_diff'](orig_file, new_file)
```
Should this be `__utils__` instead of `__salt__`?

rallytime commented on salt/modules/file.py:
```diff
@@ -2038,9 +2038,7 @@ def line(path, content=None, match=None, mode=None, location=None,
-            changes_diff = ''.join(difflib.unified_diff(
-                path_content, body
-            ))
+            changes_diff = __salt__['stringutils.get_diff'](path_content, body)
```
Same question here as above.

rallytime commented on salt/modules/file.py:
```diff
@@ -2688,7 +2683,7 @@ def _add_content(linesep, lines=None, include_marker_start=True,
 
     if block_found:
-        diff = ''.join(difflib.unified_diff(orig_file, new_file))
+        diff = __salt__['stringutils.get_diff'](orig_file, new_file)
```
Here, too?

rallytime reviewed (commented):
It seems that `__salt__` is being used instead of `__utils__` in a couple of places. There's some lint/test errors as well.

terminalmage committed:
Skip pylint false-positives

terminalmage committed:
Fix incorrect use of __salt__ when __utils__ is needed

rallytime commented:
@terminalmage This looks like the tests will need some mocking love, due to the change to `__utils__`.

terminalmage committed:
Add stringutils.get_diff to mocks

terminalmage committed:
Only try latin-1 from get_diff instead of by default

rallytime commented:
@terminalmage Looks like one last test failure to clean up here

https://jenkinsci.saltstack.com/job/pr-kitchen-ubuntu1604-py3/job/PR-48934/4/

terminalmage committed:
Fix to_unicode test

Since latin-1 is not being automatically decoded, we need to explicitly
pass it on the test.
